tools/reorder_docids.hpp

```diff
@@ -0,0 +1,319 @@
+            emit(output_docs, posting.first);
+            emit(output_freqs, posting.second);
+        }
+
+        progress.update(1);
+        posting_list.clear();
+    }
+}
+
+auto reorder_lexicon(
+    std::string const& input_lexicon,
+    std::string const& output_lexicon,
+    gsl::span<std::uint32_t const> mapping)
+
+{
+    auto doc_buffer = Payload_Vector_Buffer::from_file(input_lexicon);
+    auto documents = Payload_Vector<std::string>(doc_buffer);
+    std::vector<std::string> reordered_documents(documents.size());
+    pisa::progress doc_reorder("Reordering documents vector", documents.size());
+    for (size_t i = 0; i < documents.size(); ++i) {
+        reordered_documents[mapping[i]] = documents[i];
+        doc_reorder.update(1);
+    }
+    encode_payload_vector(reordered_documents.begin(), reordered_documents.end()).to_file(output_lexicon);
+}
+
+auto reorder_sizes(
+    binary_collection const& input_sizes,
+    std::uint64_t num_docs,
+    gsl::span<std::uint32_t const> mapping,
+    std::string_view output_basename)
+{
+    pisa::progress progress("Reordering document sizes", num_docs);
+    auto sizes = *input_sizes.begin();
+    if (sizes.size() != num_docs) {
+        throw std::invalid_argument("Invalid sizes file");
+    }
+
+    auto size_sequence = gsl::span(sizes.begin(), sizes.size());
+    std::vector<std::uint32_t> new_sizes(num_docs);
+    for (size_t i = 0; i < num_docs; ++i) {
+        new_sizes[mapping[i]] = size_sequence[i];
+        progress.update(1);
+    }
+
+    std::ofstream output_sizes(fmt::format("{}.sizes", output_basename));
+    emit(output_sizes, new_sizes.size());
+    emit(output_sizes, new_sizes.data(), num_docs);
+}
+
+void reorder_from_mapping(
+    binary_freq_collection const& input_collection,
+    binary_collection const& input_sizes,
+    std::string output_basename,
+    std::optional<std::string> const& document_lexicon,
+    std::optional<std::string> const& reordered_document_lexicon,
+    gsl::span<std::uint32_t const> mapping)
+{
+    auto num_docs = input_collection.num_docs();
+    reorder_sizes(input_sizes, num_docs, mapping, output_basename);
+    reorder_postings(input_collection, output_basename, mapping);
+    if (document_lexicon) {
+        reorder_lexicon(*document_lexicon, *reordered_document_lexicon, mapping);
+    }
+}
+
+auto reorder_random(
+    std::string const& input_basename,
+    std::string output_basename,
+    std::optional<std::string> const& document_lexicon,
+    std::optional<std::string> const& reordered_document_lexicon,
+    std::uint64_t seed) -> int
+{
+    spdlog::info("Computing random permutation");
+    binary_freq_collection input_collection(input_basename.c_str());
+    auto num_docs = input_collection.num_docs();
+    std::vector<std::uint32_t> mapping(num_docs);
+    std::iota(mapping.begin(), mapping.end(), uint32_t());
+    std::shuffle(mapping.begin(), mapping.end(), std::mt19937{seed});
```

Comment:
Travis for Clang/XCode:
```
[ 54%] Building CXX object tools/CMakeFiles/reorder-docids.dir/reorder_docids.cpp.o

In file included from /Users/travis/build/pisa-engine/pisa/tools/shards.cpp:15:

/Users/travis/build/pisa-engine/pisa/tools/reorder_docids.hpp:118:63: error: non-constant-expression cannot be narrowed from type 'std::uint64_t' (aka 'unsigned long long') to 'std::__1::mersenne_twister_engine<unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>::result_type' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]

    std::shuffle(mapping.begin(), mapping.end(), std::mt19937{seed});

                                                              ^~~~

/Users/travis/build/pisa-engine/pisa/tools/reorder_docids.hpp:118:63: note: insert an explicit cast to silence this issue

    std::shuffle(mapping.begin(), mapping.end(), std::mt19937{seed});

                                                              ^~~~

                                                              static_cast<result_type>( )

1 error generated.
```

Reply:
Yeah, for some reason, `std::mt19937` takes `unsigned int`, which is stupid but whatever. I already fixed it but I know Antonio's got some comments so I'm waiting for that to push.